### PR TITLE
Implement Shor's factorization algorithm

### PR DESCRIPTION
**Pull request by raklaptudirm**
Implemented the classical version of Shor's quantum algorithm for factorizing integers.

### CODE REVIEW

**Review by appgurueu — APPROVED** [review]
I think there is an opportunity for optimization, but it's definitely fine in its current form.

**Comment by raklaptudirm** [comment]
Any suggestions?

**Comment by appgurueu** [comment]
> Any suggestions?

See my review comment above: Leverage modular arithmetics to keep your numbers small. [Example from my Lua implementation of the Miller-Rabin primality test](https://github.com/TheAlgorithms/Lua/commit/fbb50893725534b8795e430d927a384e6349692e#diff-f3f9c6855ba634d8777a1b3b8ed41452753b598d105bc1eac012354f7823c140R4-R14).

**Comment by raklaptudirm** [comment]
> > Any suggestions?
> 
> 
> 
> See my review comment above: Leverage modular arithmetics to keep your numbers small. [Example from my Lua implementation of the Miller-Rabin primality test](https://github.com/TheAlgorithms/Lua/commit/fbb50893725534b8795e430d927a384e6349692e#diff-f3f9c6855ba634d8777a1b3b8ed41452753b598d105bc1eac012354f7823c140R4-R14).

That could be an interesting idea. But my worry is that is could make the algorithm slower, i will check though.

File: Maths/ShorsAlgorithm.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * @function isValidP
+ * @description Checks if A, B, and p fulfill A^p = mB + 1.
+ * @param {BigInt} A
+ * @param {BigInt} B
+ * @param {BigInt} p
+ * @returns Whether A, B, and p fulfill A^p = mB + 1.
+ */
+function isValidP (A, B, p) {
+  // A^p = mB + 1 => A^p - 1 = 0 (mod B)
+  return (A ** p - 1n) % B === 0n
```
[the remark this answers is not in the document]
Comment (raklaptudirm) [reply Maths/ShorsAlgorithm.js:80]: Instead of subtracting, I can just check if it is 1 (mod B).

**Pull request merged**